File: networkDelayTimeDijkstra.java
import java.util.*;

class Solution {
    // Adjacency list, defined it as per the maximum number of nodes
    // But can be defined with the input size as well
    List<List<Pair<Integer, Integer>>> adj = new ArrayList<>();
    
    public void dijkstra(int[] signalReceivedAt, int source, int n) {
        PriorityQueue<Pair<Integer, Integer>> pq = new PriorityQueue<>(
            Comparator.comparing(Pair::getKey)
        );
        pq.add(new Pair<>(0, source));
        
        // Time for starting node is 0
        signalReceivedAt[source] = 0;
        
        while (!pq.isEmpty()) {
            int currNodeTime = pq.peek().getKey();
            int currNode = pq.peek().getValue(); 
            pq.poll();
            
            if (currNodeTime > signalReceivedAt[currNode]) {
                continue;
            }
            
            // Broadcast the signal to adjacent nodes
            for (Pair<Integer, Integer> edge : adj.get(currNode)) {
                int time = edge.getKey();
                int neighborNode = edge.getValue();
                
                // Fastest signal time for neighborNode so far
                // signalReceivedAt[currNode] + time : 
                // time when signal reaches neighborNode
                if (signalReceivedAt[neighborNode] > currNodeTime + time) {
                    signalReceivedAt[neighborNode] = currNodeTime + time;
                    pq.add(new Pair<>(signalReceivedAt[neighborNode], neighborNode));
                }
            }
        }
    }
    
    public int networkDelayTime(int[][] times, int n, int k) {
        // Build the adjacency list
        for (int i = 0; i <= n; i++) {
            adj.add(new ArrayList<>());
        }
        for (int[] time : times) {
            int source = time[0];
            int dest = time[1];
            int travelTime = time[2];
            
            adj.get(source).add(new Pair<>(travelTime, dest));
        }
        
        int[] signalReceivedAt = new int[n + 1];
        Arrays.fill(signalReceivedAt, Integer.MAX_VALUE);
        dijkstra(signalReceivedAt, k, n);
        
        int answer = Integer.MIN_VALUE;
        for (int i = 1; i <= n; i++) {
            answer = Math.max(answer, signalReceivedAt[i]);
        }
        
        // INT_MAX signifies atleat one node is unreachable
        return answer == Integer.MAX_VALUE ? -1 : answer;
    }
}

File: networkDelayTimeDijkstra.py
import heapq
from typing import List, Tuple

class Solution:
    def __init__(self):
        # Adjacency list, defined it as per the maximum number of nodes
        # But can be defined with the input size as well
        self.adj = [[] for _ in range(101)]
        
    def dijkstra(self, signalReceivedAt: List[int], source: int, n: int) -> None:
        pq = [(0, source)]
        heapq.heapify(pq)
        
        # Time for starting node is 0
        signalReceivedAt[source] = 0
        
        while pq:
            currNodeTime, currNode = heapq.heappop(pq)
            
            if currNodeTime > signalReceivedAt[currNode]:
                continue
            
            # Broadcast the signal to adjacent nodes
            for time, neighborNode in self.adj[currNode]:
                # Fastest signal time for neighborNode so far
                # signalReceivedAt[currNode] + time : 
                # time when signal reaches neighborNode
                if signalReceivedAt[neighborNode] > currNodeTime + time:
                    signalReceivedAt[neighborNode] = currNodeTime + time
                    heapq.heappush(pq, (signalReceivedAt[neighborNode], neighborNode))
        
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        # Build the adjacency list
        for source, dest, travelTime in times:
            self.adj[source].append((travelTime, dest))
        
        signalReceivedAt = [float('inf')] * (n + 1)
        self.dijkstra(signalReceivedAt, k, n)
        
        answer = max(signalReceivedAt[1:])
        
        # float('inf') signifies at least one node is unreachable
        return -1 if answer == float('inf') else answer
